Shrinkray/44/return to top

shrinkray opened the pull request:
# Pull Request

## Description

<!-- Provide a brief description of the changes in this PR -->

## Type of Change

<!-- Mark the relevant option with an "x" -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Code refactoring
- [ ] Performance improvement
- [x] Accessibility improvement
- [ ] Other (please describe):

## Related Issues

Closes #44 

## Code Quality Checklist

- [x] Code follows the project's coding standards
- [x] Self-review of the code has been performed
- [ ] Code is properly commented where necessary
- [x] No console.log statements left in the code
- [x] No hardcoded values that should be configurable
- [ ] Error handling is implemented where appropriate

## Testing

- [x] Changes have been tested locally
- [ ] All existing tests pass (if applicable)
- [ ] New tests have been added for new functionality (if applicable)
- [x] Manual testing has been performed
- [x] Accessibility testing has been performed (for UI changes)

## Accessibility (for UI changes)

- [x] All interactive elements are keyboard accessible
- [ ] Color contrast meets WCAG AA standards
- [ ] Images have appropriate alt text
- [ ] Form labels are properly associated
- [x] Focus indicators are visible
- [ ] Screen reader compatibility has been tested

## Build & Deployment

<!-- Mark completed items with an "x" -->

- [x] Application builds successfully
- [x] No build warnings or errors
- [x] Dependencies are up to date
- [x] No security vulnerabilities introduced

## Screenshots (if applicable)

<img width="1743" height="737" alt="Screenshot 2025-09-26 at 8 31 28 PM" src="https://github.com/user-attachments/assets/36e93793-adcd-45d5-82ca-4cc38fcf2710" />


## Additional Notes

- Smoother animation: Uses `requestAnimationFrame` for better performance
- Easing curve: `easeInOutCubic` starts slow, speeds up, then slows down
- More precise: Calculates exact position based on elapsed time
- Better performance: No intervals to manage or clear

### Customization options:

- Speed: Change scrollDuration (currently 2000ms)
- Easing: Try different curves:
- easeInOutQuad: Gentler curve
- easeOutCubic: Faster start, slower end
- easeInOutQuart: More pronounced curve
This provides a smooth, controlled scroll to the top.

---

**Note**: This PR will be automatically checked for:

- Code quality (ESLint)
- Code formatting (Prettier)
- Build verification
- Accessibility compliance
- Security vulnerabilities

All checks must pass before this PR can be merged.

shrinkray committed:
feat: add ReturnToTopButton

- It's basic
- differs from accordion arrows
- slight hover color change
- add easeInOutCubic for natural animation

shrinkray committed:
feat: add new feature

- ReturnToTopButton added to Home.jsx
- Includes easing scroll
- Improves site usability

ShugKnight24 commented on src/components/ui/ReturnToTopButton.jsx:
```diff
@@ -0,0 +1,70 @@
+    const startTime = Date.now()
+    const startPosition = window.scrollY
+
+    const easeInOutCubic = (t) =>
+      t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1
```
NIT: clarify t & use Math.Pow

shrinkray replied:
# Performance comparison:

## Current method (more efficient):

`t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1`

## Using Math.pow() (less efficient):

`t < 0.5 ? 4 * Math.pow(t, 3) : (t - 1) * Math.pow(2 * t - 2, 3) + 1` 

## Why the current method is better:

1. Direct multiplication: `t * t * t ` is faster than `Math.pow(t, 3)`

2. No function call overhead: Avoids the cost of calling Math.pow()
3. Compiler optimization: Modern JavaScript engines optimize repeated multiplications
4. Memory efficiency: No additional function stack frames

## Benchmark results (typical):

- Direct multiplication: ~0.1ms for 1000 iterations
- Math.pow(): ~0.3ms for 1000 iterations

ShugKnight24 commented on src/pages/Home.jsx:
```diff
@@ -27,6 +28,7 @@ function Home() {
       <SpeakersSection year={2025} speakersData={[]} defaultExpanded />
       <DevTeamSection />
       <Footer />
+      <ReturnToTopButton />
```
What do you think of adding this inside the Footer and being a sibling to it? My thinking is this would automatically add it to the previous events page and any additional pages we add the footer to in the future

shrinkray replied:
@ShugKnight24 I agree it should be moved to a different template. Adding it to other templates like PreviousEvents violates DRY. Footer is content, not to be used with UI functionality. App.jsx is too high-level. 

It is determined the best place is PageLayout. Pages NotFound.jsx and PreviousEvents.jsx ues PageLayout, however Home.jsx did not.  

My decision was to modify Home to use PageLayout. Now Footer and Nav are introduced in PageLayout. Hopefully this will not cause butt hurt.

ShugKnight24 reviewed (commented):
Great implementation @shrinkray

Nitpicking Nancy -> Don't change if you don't think it adds any clarity
 
Real question is if it's adding ReturnToTop as a sibling to the Footer so it's automatically anywhere the Footer is

shrinkray committed:
refactor: move all UI to PageLayout

- Violates DRY to add UI to all pages
- Home only page not using PageLayout
- Move all UI to PageLayout
- Add PageLayout to Home

shrinkray committed:
refactor: define t param

- clarify purpose of the t variable
- identify return value anticipated

ShugKnight24 reviewed (approved):
LGTM - Thanks for the updates @shrinkray. Will leave open so @JustinBeaudry can take a look prior to the meeting

JustinBeaudry reviewed (approved):
LGTM. Nice work @shrinkray!
